### ISSUE TITLE
Creating a request and not executing it directly causes uncaught TypeError: Cannot read property 'length' of undefined

### ISSUE THREAD

**Issue by themaarten** [visionmedia/superagent#509]
When

1. using the Node version of Superagent (v0.21.0 in my case),
2. creating a superagent request `r`,
3. calling the `r.request()` function, which will happen indirectly by (amongst others) setting a header through `r.set()` or data through `r.send()`, and
4. executing the request with a delay of more than (in my case) 60 seconds, by not calling calling the `.end()` immediately,

then superagent will throw an uncaught exception.

```
/some/path/node_modules/superagent/lib/node/index.js:746
  if (2 == fn.length) return fn(err, res);
             ^
TypeError: Cannot read property 'length' of undefined
    at Request.callback (/some/path/node_modules/superagent/lib/node/index.js:746:14)
    at ClientRequest.<anonymous> (/some/path/node_modules/superagent/lib/node/index.js:711:10)
    at ClientRequest.emit (events.js:95:17)
    at Socket.socketOnEnd [as onend] (http.js:1568:9)
    at Socket.g (events.js:180:16)
    at Socket.emit (events.js:117:20)
    at _stream_readable.js:929:16
    at process._tickCallback (node.js:419:13)
```

This is whats happening: after being idle for some time, the http(s).ClientRequest emits an error event:
```{ [Error: socket hang up] code: 'ECONNRESET' }```
Superagent is trying to pass the error on to the callback which has not been registered through a call to `.end()`.

I have to admit that creating a request and not calling `.end()` directly may be an extravagant use case. I only ran into the error when trying out some requests through the REPL. On the other hand, more people seem to run into it. See the comments on issue #339.

It should either be discouraged in the documentation to prepare a request and and execute (i.e. call `.end()`) later, _or_ the `http(s).ClientRequest` should not be created until `.end()` is called.

**Comment by kellyrmilligan** [comment]
is it not possible to have a request stored in a var and then executing it when you call .end with superagent?  I'm trying to build up the headers and params in separate methods for a wrapper i'm creating....

**Comment by lucastschmidt** [comment]
Humm, I am having this error in my app, using this exact same version 0.21.0. 

In my case, I was override the end method, and doing some extra checks when the user session expired with an API. If the checks took too long, it would break the app.

**Comment by focusaurus** [comment]
I think this has been fixed already in v1.6.1. I am unable to reproduce there and I don't see any calls to `this.request()` during `req.set()` or `req.send()`. @themaarten if you could confirm that would be great. Please re-open if this is still an issue on the latest release.

**Comment by themaarten** [comment]
I just did a test with v1.6.1. The issue seems to have been fixed indeed!
